src/Map.ts

```diff
@@ -0,0 +1,565 @@
+import { Applicative, Applicative1, Applicative2, Applicative2C, Applicative3, Applicative3C } from './Applicative'
+import { Separated } from './Compactable'
+import { Either } from './Either'
+import { Foldable, Foldable1, Foldable2, Foldable3 } from './Foldable'
+import { Predicate, tuple, Refinement } from './function'
+import { HKT, Type, Type2, Type3, URIS, URIS2, URIS3 } from './HKT'
+import { getMapMonoid, Monoid } from './Monoid'
+import { Option, none, some } from './Option'
+import { Setoid, fromEquals } from './Setoid'
+import { Unfoldable } from './Unfoldable'
+import { Semigroup } from './Semigroup'
+
+/**
+ * Calculate the number of key/value pairs in a Map
+ *
+ * @since 1.14.0
+ */
+export const size = <K, A>(d: Map<K, A>): number => {
+  return d.size
+}
+
+/**
+ * Test whether a Map is empty
+ *
+ * @since 1.14.0
+ */
+export const isEmpty = <K, A>(d: Map<K, A>): boolean => {
+  return d.size === 0
+}
+
+/**
+ * @since 1.14.0
+ */
+export const keys = <K, A>(d: Map<K, A>): Array<K> => {
+  return Array.from(d.keys())
+}
+
+/**
+ * @since 1.14.0
+ */
+export const unsafeLookup = <K, A>(k: K, d: Map<K, A>): A => {
+  return d.get(k)!
+}
+
+/**
+ * @since 1.14.0
+ */
+export function collect<K, A, B>(d: Map<K, A>, f: (k: K, a: A) => B): Array<B> {
+  const out: Array<B> = []
+  const ks = keys(d).sort()
+  for (let key of ks) {
+    const lookup = d.get(key)
+    if (lookup !== undefined) {
+      out.push(f(key, lookup))
+    }
+  }
+  return out
+}
+
+/**
+ * @since 1.14.0
+ */
+export function toArray<K, A>(d: Map<K, A>): Array<[K, A]> {
+  return collect(d, (k, a: A) => tuple(k, a))
+}
+
+/**
+ * Unfolds a Map into a list of key/value pairs
+ *
+ * @since 1.14.0
+ */
+export const toUnfoldable = <F>(unfoldable: Unfoldable<F>) => <K, A>(d: Map<K, A>): HKT<F, [K, A]> => {
+  const arr = toArray(d)
+  const len = arr.length
+  return unfoldable.unfoldr(0, b => (b < len ? some(tuple(arr[b], b + 1)) : none))
+}
+
+/**
+ * Insert or replace a key/value pair in a Map
+ *
+ * @since 1.14.0
+ */
+export function insert<KS, K extends KS, A>(k: K, a: A, d: Map<KS, A>): Map<KS | K, A>
+export function insert<K, A>(k: K, a: A, d: Map<K, A>): Map<K, A> {
+  const m = new Map(d)
+  m.set(k, a)
```

Comment:
Please try to be conservative and return the original Map if the operation is a noop (value already present at that key, tested using strict equality)

Reply:
Hmmm. So, you want me to always perform a `get` first to check the value?

Reply:
@joshburgess yes